src/Domain/CoordinateMaps/CylindricalEndcap.cpp

```diff
@@ -0,0 +1,914 @@
+             "and the map has not been tested for this choice of parameters.");
+
+  const double dist_spheres = sqrt(square(center_one_[0] - center_two_[0]) +
+                                   square(center_one_[1] - center_two_[1]) +
+                                   square(center_one_[2] - center_two_[2]));
+  ASSERT(dist_spheres + radius_one < radius_two,
+         "CylindricalEndcap: The map has been tested only for the case when "
+         "sphere_one is contained inside sphere_two");
+
+  const double proj_radius_two = sqrt(square(center_two_[0] - proj_center_[0]) +
+                                      square(center_two_[1] - proj_center_[1]) +
+                                      square(center_two_[2] - proj_center_[2]));
+  ASSERT(proj_radius_two < radius_two_,
+         "CylindricalEndcap: The map has been tested only for the case when "
+         "proj_center is contained inside sphere_two");
+
+  // Check if we are too close to singular.
+  ASSERT(acos(abs(cos_alpha)) < abs(0.95 * asin(cos_theta)),
+         "CylindricalEndcap: Parameters are close to where the map becomes "
+         "non-invertible.  The map has not been tested for this case.");
+
+  // Check if opening angle is small enough.
+  const double max_opening_angle = M_PI / 3.0;
+  const double max_proj_center_x =
+      x_plane_ -
+      radius_one_ * sqrt(1.0 - square(cos_theta)) / tan(max_opening_angle);
+  ASSERT(proj_center_[0] < max_proj_center_x,
+         "CylindricalEndcap: proj_center is too close to x_plane. The "
+         "map has not been tested for this case.");
+  const double tan_beta = sqrt(square(center_one_[1] - proj_center_[1]) +
+                               square(center_one_[2] - proj_center_[2])) /
+                          (max_proj_center_x - proj_center_[0]);
+  ASSERT(tan_beta < tan(max_opening_angle),
+         "CylindricalEndcap: opening angle is too large. The map has not "
+         "been tested for this case.");
+#endif
+}
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> CylindricalEndcap::operator()(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  CylindricalEndcap_detail::InnerSphereMap inner_map(center_one_, radius_one_,
+                                                     x_plane_);
+
+  // lower_coords are the mapped coords on the surface of sphere 1.
+  const std::array<return_type, 3> lower_coords = inner_map(source_coords);
+
+  // upper_coords are the mapped coords on the surface of sphere 2.
+  const auto lambda = CylindricalEndcap_detail::scale_factor<return_type>(
+      lower_coords, proj_center_, center_two_, radius_two_);
+
+  std::array<return_type, 3> upper_coords = lower_coords;
+  for (size_t i = 0; i < 3; ++i) {
+    gsl::at(upper_coords, i) =
+        gsl::at(proj_center_, i) +
+        (gsl::at(lower_coords, i) - gsl::at(proj_center_, i)) * lambda;
+  }
+
+  std::array<return_type, 3> crap = lower_coords;
```

Comment:
Do these end up being used?

Reply:
Rats. No they are not used and I am removing them.  I was using them for debugging (hence the names, which were supposed to remind me to remove them).